pyacq/core/stream/ringbuffer.py

```diff
@@ -0,0 +1,304 @@
+                            (data.dtype, self.dtype))
+        
+        # by default, index advances by the size of the chunk
+        if index is None:
+            index = self._write_index + dsize
+        
+        assert dsize <= index - self._write_index, ("Data size is %d, but index "
+                                                    "only advanced by %d." % 
+                                                    (dsize, index-self._write_index)) 
+
+        revert_inds = [self._read_index, self._write_index]
+        try:
+            # advance write index. This immediately prevents other processes from
+            # accessing memory that is about to be overwritten.
+            self._set_write_index(index)
+            
+            # decide if any skipped data needs to be filled in
+            fill_start = max(self._read_index + 1, self._write_index + 1 - bsize)
+            fill_stop = self._write_index + 1 - dsize
+            
+            if fill_stop > fill_start:
+                # data was skipped; fill in missing regions with 0 or nan.
+                self._write(fill_start, fill_stop, self._filler)
+                revert_inds[1] = fill_stop-1
+                
+            self._write(self._write_index + 1 - dsize, self._write_index + 1, data)
+                
+            self._set_read_index(index)
+        except:
+            # If there is a failure writing data, revert read/write pointers
+            self._set_read_index(revert_inds[0])
+            self._set_write_index(revert_inds[1])
+            raise
+
+    def _write(self, start, stop, value):
+        # get starting index
+        bsize = self.shape[0]
+        dsize = stop - start
+        i = start % bsize
+        
+        if self.double:
+            self.buffer[i:i+dsize] = value
+            i += bsize
+        
+        if i + dsize <= self.buffer.shape[0]:
+            self.buffer[i:i+dsize] = value
+        else:
+            n = self.buffer.shape[0]-i
+            self.buffer[i:] = value[:n]
+            self.buffer[:dsize-n] = value[n:]
+        
+    def __getitem__(self, item):
+        if isinstance(item, tuple):
+            first = item[0]
+            rest = (slice(None),) + item[1:]
+        else:
+            first = item
+            rest = None
+        
+        if isinstance(first, (int, np.integer)):
+            start = self._interpret_index(first)
+            stop = start + 1
+            data = self.get_data(start, stop)[0]
+            if rest is not None:
+                data = data[rest[1:]]
+        elif isinstance(first, slice):
+            start, stop, step = self._interpret_index(first)
+            data = self.get_data(start, stop)[::step]
+            if rest is not None:
+                data = data[rest]
+        else:
+            raise TypeError("Invalid index type %s" % type(first))
+        
+        return data
+
+    def get_data(self, start, stop, copy=False, join=True):
```

Comment:
I more often need get_data from stop and  lentgh.
This was is ex-behavior with_get_array_slice.
So I very often do this ```get_data(stop-length, lentgh)```.
I have no opinion on that but my intuition lead to prefer have ```get_data(stop, length)```  directly. Even if it sounds more natural to have ```get_data-stop, start)```